Improve SQLAlchemy store parsing and documentation

drewmcdonald opened the pull request:
## What changes are proposed in this pull request?
 
Addresses #1266; enables users to specify database drivers in a SQLAlchemy URI. Without this change, a URI beginning with `database+driver://...` is not recognized as a database at all, and MLflow reverts to the file store.

- [x] enable user-specified drivers in database URIs
- [x] confirm that db_type is supported; raise error instead of reverting to file store
- [x] validate db driver string if specified
- [x] improve documentation on SQLAlchemy connection strings, linking to docs
  
## How is this patch tested?
 
Added tests to [test_sqlalchemy_store.py](https://github.com/mlflow/mlflow/blob/7e6a18af6386bf8462507ceb9c7f03ece89039da/tests/store/test_sqlalchemy_store.py#L33).

## Release Notes
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Allow user-specified drivers in SQLAlchemy database URIs. Additionally, throw an error if attempting to use a bad or unsupported database specification instead of silently reverting to the file store.
  
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [x] Docs
- [x] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

drewmcdonald committed:
scrub '+{driver}' from sqlalchemy URI when identifying db_type

mparkhe commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -55,15 +55,15 @@ def __init__(self, db_uri, default_artifact_root):
+                       ``db_type+driver://<user_name>:<password>@<host>:<port>/<database_name>`
                        Supported database types are ``mysql``, ``mssql``, ``sqlite``,
-                       and ``postgresql``.
+                       and ``postgresql``. The driver element is optional.
```
Can we give some context for new users. Perhaps add a link to this [SQLAlchemy Engine Configuration](https://docs.sqlalchemy.org/en/13/core/engines.html#database-urls) page you pointed to in issue #1266. The page has a lot of specific examples as well, which is great!

mparkhe commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -55,15 +55,15 @@ def __init__(self, db_uri, default_artifact_root):
         super(SqlAlchemyStore, self).__init__()
         self.db_uri = db_uri
-        self.db_type = urllib.parse.urlparse(db_uri).scheme
+        self.db_type = urllib.parse.urlparse(db_uri).scheme.split('+')[0]
```
If `driver` is specified, can we add some validation to what is acceptable in `driver` string? Something like this

```
_BAD_DRIVER_STRING_MSG = "DBAPI driver name can only include alphanumerics, underscores (_), and dashes (-)"
_VALID_DRIVER_STRING = re.compile(r"[/\w\-_]*$")
def _validate_driver(driver):
  # validates DBAPI driver name used in db_uri parameter
  if not _VALID_DRIVER_STRING.match(driver):
    raise MlflowException("Invalid DBAPI driver string: '%s'. %s" % (driver, _BAD_DRIVER_STRING_MSG), INVALID_PARAMETER_VALUE)
```

Similar examples in [mlflow.utils.validation](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/validation.py)

mparkhe reviewed (changes requested):
Thanks. This is a great change. Couple of comments.

Would also appreciate if you could add tests to ensure valid `driver` string in `db_uri` parameter.

drewmcdonald commented:
Thanks for the feedback! Since you all seem interested in fleshing this out a little further I'll take a few days and go for some deeper checks and changes. Mainly it feels to me like this area is too prone to silently reverting to a file store. 

Here's a combined TODO. Anything else in this realm worth adding?

- [ ] validate db driver string if specified
- [ ] improve documentation on sqlalchemy connection strings, linking to SQA docs
- [ ] test that db_type is supported

mparkhe commented:
Hi @drewmcdonald:

That sounds right. I would add one more check

- [ ] Does the code correctly handle input uri strings like `"mysql+driver1+driver2://..."` 

The current approach of `scheme.split('+')[0]` would ignore the second driver. Is that legit syntax? Can connection string have multiple JDBC drivers? Error handling if not. Else, each driver needs to be validated.

---

This is a good change. You may have seen we sent out [an RC advisory](https://mlflow.org/news/2019/05/22/1.0.0-release-candidate-1/index.html) yesterday. This would be a good one to get into MLflow 1.0. Please lemme know if you will have time to address it in the next couple of days.

drewmcdonald committed:
enhancements to db uri parsing in SQLAlchemy store
- allow dialect+driver://... scheme format
- raise error if db type is unsupported
- raise error if db driver is not believable

drewmcdonald committed:
missing space around ' = '

drewmcdonald committed:
improve documentation on sqlalchemy connection strings/db_uris

drewmcdonald commented:
@mparkhe I think this is ready for another review

Commit:
minor fixes, refactor, and unit tests

Commit:
fixes

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into enable_sql_drivers

stbof commented on docs/source/tracking.rst:
```diff
@@ -71,7 +71,9 @@ call :py:func:`mlflow.set_tracking_uri`.
+- Database encoded as ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``.
+Mlflow supports the dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. For more
+details refer `SQLAlchemy database uri <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_
 - HTTP server (specified as ``https://my-server:5000``), which is a server hosting an :ref:`MLFlow tracking server <tracking_server>`.
```
For more details refer SQLAlchemy database uri -> For more details, see SQLAlchemy database URI

stbof commented on docs/source/tracking.rst:
```diff
@@ -265,12 +267,13 @@ The backend store is where MLflow Tracking Server stores experiment and run meta
+Use ``--backend-store-uri`` to configure the type of backend store. A *file store* backend is
+specified as ``./path_to_store`` or ``file:/path_to_store``., while a *database-backed store* is
+specified with a `SQLAlchemy database URI <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_.
+The database URI typically takes the format ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``.
```
A *file store* backend is
specified as ``./path_to_store`` or ``file:/path_to_store``., while a *database-backed store* is
specified with a `SQLAlchemy database URI ->

You specify a *file store* backend  as ``./path_to_store`` or ``file:/path_to_store`` and a *database-backed store* as a `SQLAlchemy database URI

stbof commented on docs/source/tracking.rst:
```diff
@@ -265,12 +267,13 @@ The backend store is where MLflow Tracking Server stores experiment and run meta
+The database URI typically takes the format ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``.
+MLflow supports the database dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. Drivers
+are optional; SQLAlchemy will use a dialect's default driver if a driver is not specified. For
+backwards compatibility, ``--file-store`` is an alias for ``--backend-store-uri``.
```
SQLAlchemy will use a dialect's default driver if a driver is not specified. -> 

If you do not specify a driver, SQLAlchemy uses a dialect's default driver.

stbof commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -19,20 +19,52 @@
-    interacts with SQL store using SQLAlchemy abstractions defined for MLflow entities.
+    SQLAlchemy compliant backend store for tracking meta data for MLflow entities. Currently,
+    MLflow supports the database dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``.
+    As specified in the
```
Currently, MLflow supports the database dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. ->

MLflow supports the database dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``.

stbof commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -19,20 +19,52 @@
+    `SQLAlchemy docs <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_ ,
+    the database URI is expected in the format
+    ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``. SQLAlchemy will use
+    a dialect's default driver if a driver is not specified.
```
SQLAlchemy will use a dialect's default driver if a driver is not specified. -> 

If you do not specify a driver, SQLAlchemy uses a dialect's default driver

stbof commented on mlflow/utils/validation.py:
```diff
@@ -30,6 +31,10 @@
 
+_VALID_DRIVER_STRING = re.compile(r"^[a-z0-9_]*$")
+_BAD_DRIVER_STRING_MSG = "Driver can only include lowercase alphanumerics and underscores (_)"
+_UNSUPPORTED_DB_TYPE_MSG = "Supported database engines are {%s}" % ', '.join(DATABASE_ENGINES)
```
Driver can only include lowercase alphanumerics and underscore -> Driver name can include only lowercase alphanumerics and underscores

Commit:
review comments

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into enable_sql_drivers

dbczumar commented on mlflow/utils/validation.py:
```diff
@@ -30,6 +31,10 @@
 MAX_ENTITY_KEY_LENGTH = 250
 
+_VALID_DRIVER_STRING = re.compile(r"^[a-z0-9_]*$")
+_BAD_DRIVER_STRING_MSG = "Driver name can include only lowercase alphanumerics and underscores (_)"
```
Does this restriction come from SQLAlchemy or a particular RFC? If not, are there any disadvantages to removing this regex matching check?

drewmcdonald replied:
Come to think of it, no. I think we'd be fine to drop it. SQLAlchemy raises an error on invalid drivers.

```python
from sqlalchemy import create_engine

eng_noexist = create_engine('mysql+blahblah://...')
# --> NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:mysql.blahblah

eng_badchar = create_engine('mysql+blah*blah://...')
# --> ArgumentError: Could not parse rfc1738 URL from string 'mysql+blah*blah://...'

```

dbczumar reviewed (approved):
LGTM. Left one small comment about driver validation.

drewmcdonald committed:
leave driver validation to SQLAlchemy

drewmcdonald commented:
@dbczumar and @mparkhe see last commit that removes unnecessary driver validation since SQLAlchemy seems to handle those well enough already:

```python
from sqlalchemy import create_engine

eng_noexist = create_engine('mysql+blahblah://...')
# --> NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:mysql.blahblah

eng_badchar = create_engine('mysql+blah*blah://...')
# --> ArgumentError: Could not parse rfc1738 URL from string 'mysql+blah*blah://...'
```

Feel free to rollback if you'd rather.

mparkhe commented:
> @dbczumar and @mparkhe see last commit that removes unnecessary driver validation since SQLAlchemy seems to handle those well enough already:
> 
> ```python
> from sqlalchemy import create_engine
> 
> eng_noexist = create_engine('mysql+blahblah://...')
> # --> NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:mysql.blahblah
> 
> eng_badchar = create_engine('mysql+blah*blah://...')
> # --> ArgumentError: Could not parse rfc1738 URL from string 'mysql+blah*blah://...'
> ```
> 
> Feel free to rollback if you'd rather.

Code changes look good! Once tests complete, I can merge this in. Thanks @drewmcdonald for teh quick turn around on this.

drewmcdonald commented:
@mparkhe and @dbczumar we missed a usage of `urllib.parse.urlparse(uri).scheme` in `mlflow.tracking.utils`.

https://github.com/mlflow/mlflow/blob/47b6575bf56d9aa06c2d8fe988a25e05326283fb/mlflow/tracking/utils.py#L96-L99

This raises an error when called by `_get_store` in `mlflow.server.handlers`. As a result, `mlflow server --backend-store-uri=db_type+driver://...` is still not operational and user-specified drivers are not supported.

How do you want to approach this? Happy to be helpful.
